noxfile.py

```diff
@@ -17,20 +24,207 @@
     "clang-format",
 ]
 
+#
+# Helpers for use within this file.
+#
+
+
+def simulator_support_matrix() -> List[Tuple[str, str, str]]:
+    """
+    Get a list of supported simulator/toplevel-language/GPI-interface tuples.
+    """
+
+    # Simulators with support for VHDL through VHPI, and Verilog through VPI.
+    standard = [
+        (sim, toplevel_lang, gpi_interface)
+        for sim in ("activehdl", "rivierapro", "xcelium")
+        for toplevel_lang in ("verilog", "vhdl")
+        for gpi_interface in ("vpi", "vhpi")
+        if (toplevel_lang, gpi_interface) in (("verilog", "vpi"), ("vhdl", "vhpi"))
+    ]
+
+    # Special-case simulators.
+    special = [
+        ("cvc", "verilog", "vpi"),
+        ("ghdl", "vhdl", "vpi"),
+        ("icarus", "verilog", "vpi"),
+        ("questa", "verilog", "vpi"),
+        ("questa", "vhdl", "fli"),
+        ("questa", "vhdl", "vhpi"),
+        ("verilator", "verilog", "vpi"),
+        ("vcs", "verilog", "vpi"),
+    ]
+
+    return standard + special
+
+
+def env_vars_for_test(
+    sim: Optional[str], toplevel_lang: Optional[str], gpi_interface: Optional[str]
+) -> Dict[str, str]:
+    """Prepare the environment variables controlling the test run."""
+    e = {}
+    if sim is not None:
+        e["SIM"] = sim
+    if toplevel_lang is not None:
+        e["TOPLEVEL_LANG"] = toplevel_lang
+    assert not (toplevel_lang == "verilog" and gpi_interface != "vpi")
+    if toplevel_lang == "vhdl" and gpi_interface is not None:
+        e["VHDL_GPI_INTERFACE"] = gpi_interface
+
+    return e
+
+
+def stringify_dict(d: Dict[str, str]) -> str:
+    return ", ".join(f"{k}={v}" for k, v in d.items())
+
+
+#
+# Development pipeline
+#
+# - Use nox to build an sdist; no separate build step is required.
+# - Run tests against the installed sdist.
+# - Collect coverage.
+#
+
 
 @nox.session
-def tests(session: nox.Session) -> None:
-    """run cocotb regression suite"""
+def dev_build(session: nox.Session) -> None:
+    session.warn("No building is necessary for development sessions.")
+
+
+@nox.session
+def dev_test(session: nox.Session) -> None:
+    """Run all development tests as configured through environment variables."""
+
+    dev_test_nosim(session)
+    dev_test_sim(session, sim=None, toplevel_lang=None, gpi_interface=None)
+
+
+@nox.session
+@nox.parametrize("sim,toplevel_lang,gpi_interface", simulator_support_matrix())
+def dev_test_sim(
+    session: nox.Session,
+    sim: Optional[str],
+    toplevel_lang: Optional[str],
+    gpi_interface: Optional[str],
+) -> None:
+    """Test a development version of cocotb against a simulator."""
+
     session.env["CFLAGS"] = "-Werror -Wno-deprecated-declarations -g --coverage"
     session.env["COCOTB_LIBRARY_COVERAGE"] = "1"
     session.env["CXXFLAGS"] = "-Werror"
     session.env["LDFLAGS"] = "--coverage"
-    session.install(*test_deps)
+
+    session.run("pip", "install", *test_deps, *coverage_deps)
```

Comment:
This is cute, but I'm not a fan. `--no-install` was really only ever intended as an additional speed-up for `-R` (reusing session) not as a way to conditionally skip certain installs. That should probably be handled with posargs and an argparser.

Reply:
I removed this part.